### PR TITLE
[UNR-5790] fix actor going from dormant to awake

### PR DESCRIPTION
**Pull request by simonsarginson**
#### Description
fix actor going from dormant to awake

#### Tests
Covered by automated tests.

GDK: https://github.com/spatialos/UnrealGDK/pull/3279
426: https://github.com/improbableio/UnrealEngine/pull/1028
425: https://github.com/improbableio/UnrealEngine/pull/1029

### CODE REVIEW

File: RequireSetup [hunk 1]
```diff
@@ -1,4 +1,4 @@
 Increment the below number whenever it is required to run Setup.bat or Setup.sh as part of a new commit.
 Our git hooks will detect this file has been updated and automatically run Setup.bat on pull.
 
-94
+95
```
Comment (simonsarginson) [inline RequireSetup:4]:
```suggestion
95

```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 2]
```diff
@@ -365,16 +366,16 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
 	UFUNCTION(BlueprintPure, Category = "Spatial Functional Test")
 	bool WasLoadedFromTakenSnapshot();
 
-	template <typename T>
-	static int GetNumberOfActorsOfType(UWorld* World)
+	template <typename Actor_Type>
+	static SIZE_T CountActors(UWorld* World)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]: drive by

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 3]
```diff
@@ -365,16 +366,16 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
...
 	{
-		int Counter = 0;
-		for (TActorIterator<T> Iter(World); Iter; ++Iter)
-		{
-			Counter++;
-		}
+		return CountActors<Actor_Type>(World, [](Actor_Type*){ return true; });;
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:372]: Nit
```suggestion
		return CountActors<Actor_Type>(World, [](Actor_Type*){ return true; });
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 4]
```diff
@@ -153,6 +153,16 @@ void ActorSystem::ProcessUpdates(const FEntitySubViewUpdate& SubViewUpdate)
 				ApplyComponentAdd(Delta.EntityId, Change.ComponentId, Change.Data);
 				ComponentAdded(Delta.EntityId, Change.ComponentId, Change.Data, ToResolveOps);
 			}
+			for (const ComponentChange& Change : Delta.ComponentsRemoved)
+			{
+				// NOTE: We have to deal the with dormant before ComponentUpdated
+				// otherwise a SubViewUpdate that contains a component update that
+				// also removes the dormant component will have its updates ignored.
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: I don't fully understand this what else can remove the dormant component? Can we do all component removes before updates or only this one?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: Basically the dormant component still being there causes the update to be ignored which normally is fine but is a race if both the dormant component are removed and updates occur.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 2 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]: It's `SIZE_T` due to `Algo::Count` returning `SIZE_T`, right? Just clarifying as I don't really see it elsewhere.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 5]
```diff
@@ -365,16 +366,16 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
 	UFUNCTION(BlueprintPure, Category = "Spatial Functional Test")
 	bool WasLoadedFromTakenSnapshot();
 
-	template <typename T>
-	static int GetNumberOfActorsOfType(UWorld* World)
+	template <typename Actor_Type>
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:369]: Nit, could also be `TActorType`
```suggestion
	template <typename ActorType>
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 6]
```diff
@@ -153,6 +153,16 @@ void ActorSystem::ProcessUpdates(const FEntitySubViewUpdate& SubViewUpdate)
 				ApplyComponentAdd(Delta.EntityId, Change.ComponentId, Change.Data);
 				ComponentAdded(Delta.EntityId, Change.ComponentId, Change.Data, ToResolveOps);
 			}
+			for (const ComponentChange& Change : Delta.ComponentsRemoved)
+			{
+				// NOTE: We have to deal the with dormant before ComponentUpdated
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:158]: So in the end you've decided to not handle this at a higher level? I'd imagined finding `DORMANT_COMPONENT_IDs` in `ComponentsRemoved` before going into this loop or something.

**Review by improbable-dmitrii — APPROVED** [review]
Thank you for fixing this!

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 4 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: In ComponentUpdated we still have no channel but the ActorSubView also does not have the SpatialConstants::DORMANT_COMPONENT_ID component. We need ComponentRemoved to make a channel for us first.
```
	USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(EntityId);
	if (Channel == nullptr)
	{
		// If there is no actor channel as a result of the actor being dormant, then assume the actor is about to become active.
		if (ActorSubView->HasComponent(EntityId, SpatialConstants::DORMANT_COMPONENT_ID))
		{
			// ...
		}
		else
		{
			// <-- we hit this
			UE_LOG(LogActorSystem, Log,
				   TEXT("Worker: %s Entity: %lld Component: %d - No actor channel for update."));
			return;
		}
	}

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 6 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:158]: We could split them in to two before that is true. Let's see if people are happy with the overall approach and if so I can add that.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 3 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:372]: Good catch.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 5 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:369]: +1

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 2 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]: yeah other places use `int32` I could change to that.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 7]
```diff
@@ -365,16 +366,16 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
...
-		for (TActorIterator<T> Iter(World); Iter; ++Iter)
-		{
-			Counter++;
-		}
+		return CountActors<Actor_Type>(World, [](Actor_Type*){ return true; });;
+	}
 
-		return Counter;
+	template <typename Actor_Type>
+	static SIZE_T CountActors(UWorld* World, TFunction<bool(Actor_Type*)> Pred)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:376]:
```suggestion
	static int32 CountActors(UWorld* World, TFunction<bool(Actor_Type*)> Pred)
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 2 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]:
```suggestion
	static int32 CountActors(UWorld* World)
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 3 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:372]:
```suggestion
		return CountActors<ActorType>(World, [](Actor_Type*){ return true; });
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 8]
```diff
@@ -365,16 +366,16 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
...
+	{
+		return Algo::CountIf(TActorRange<Actor_Type>(World),Pred);
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:378]:
```suggestion
		return Algo::CountIf(TActorRange<ActorType>(World), Pred);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 9]
```diff
@@ -859,8 +859,9 @@ bool USpatialActorChannel::ReplicateSubobject(UObject* Object, const FReplicatio
 	SCOPE_CYCLE_UOBJECT(ReplicateSubobjectSpecificClass, Object);
 
 	bool bCreatedReplicator = false;
+	bool bOutPreviouslyDormant = false;
 
-	FObjectReplicator& Replicator = FindOrCreateReplicator(Object, &bCreatedReplicator).Get();
+	FObjectReplicator& Replicator = FindOrCreateReplicator(Object, &bCreatedReplicator, &bOutPreviouslyDormant).Get();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:864]: DormantReplicatorMap is publicly available, we should only make engine changes where necessary, I'd rather we rework this so we don't need to change the engine API.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:864]:
```suggestion
	// Set here because FindOrCreateReplicator will remove the replicator from the dormancy map
	const bool bPreviouslyDormant = Connection->DormantReplicatorMap.Find(Object);
	FObjectReplicator& Replicator = FindOrCreateReplicator(Object, &bCreatedReplicator).Get();
```

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:864]: Something like this? I don't love how order dependent this is and reliant on opaque implementation details, but I can see how avoiding engine changes might also have advantages.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h [hunk 10]
```diff
@@ -365,16 +366,18 @@ class SPATIALGDKFUNCTIONALTESTS_API ASpatialFunctionalTest : public AFunctionalT
 	UFUNCTION(BlueprintPure, Category = "Spatial Functional Test")
 	bool WasLoadedFromTakenSnapshot();
 
-	template <typename T>
-	static int GetNumberOfActorsOfType(UWorld* World)
+	template <typename ActorType>
+	static int32 CountActors(UWorld* World)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]: This looks like worse performance, why change it?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp [hunk 11]
```diff
@@ -17,33 +29,35 @@ void ADynamicActorAwakeAfterDormantChangePropertyTest::PrepareTest()
...
-		TEXT("ClientRequireDormancyTestState"), FWorkerDefinition::AllClients, nullptr, nullptr,
-		[this](float DeltaTime) {
-			RequireDormancyActorCount(1);
-			RequireDormancyTestState(DORM_DormantAll, /*TestRepProperty*/ 0, /*ActorCount*/ 1);
-			FinishStep();
-		},
-		5.0f);
+	// Step 2 Require channel to actually be dormant
+	AddStep(TEXT("ServerAssertActuallyDormant"), FWorkerDefinition::Server(1), nullptr, nullptr, [this](float) {
+		const TSet<TSharedPtr<FNetworkObjectInfo>, FNetworkObjectKeyFuncs>& Actives = DormancyActor->GetNetDriver()->GetNetworkObjectList().GetActiveObjects();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp:40]: This would also be true if the object hasn't been replicated yet. What about checking the DormantObjectOnAllConnections like you had initially?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp [hunk 12]
```diff
@@ -17,33 +29,35 @@ void ADynamicActorAwakeAfterDormantChangePropertyTest::PrepareTest()
...
-		{
-			ADormancyTestActor* DormancyTestActor = *Iter;
-			DormancyTestActor->TestIntProp = 1;
-		}
+		DormancyActor->TestIntProp = 1;
 		FinishStep();
 	});
 
-	// Step 4 - Give chance for property to be replicated to clients
+	// Step 5 - Give chance for property to be replicated to clients, if we don't do this the
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp:59]: Driveby - but if we want to really ensure it's replicated we should add a DormancyActor->ForceNetUpdate() after setting it to 1 in Step 4

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp:59]: Oh - but we don't want it to actually replicated - right. Ignore me.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h (hunk 10 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Public/SpatialFunctionalTest.h:370]: Yeah its a good point I though I wanted a predicated count but it turns out I didn't so this is left over.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp (hunk 11 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/DynamicActorAwakeAfterDormantChangePropertyTest.cpp:40]: Fair point

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:864]: Ok I am coming around to this @m-samiec are you happy with the change outlined in my suggestion?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: Lovely

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:864]: Lovely

**Review by m-samiec — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 4 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: Can we assume that the existence of the update means that it's no longer dormant? Or that we create a channel for doing them?

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: Thinking about how to make this do something easier I think it turns it into a bigger PR. It feels like a case of having the spatial and unreal stuff very intertwined in their assumptions. We should probably make dormancy a subview filter but I'm wondering if there is ever a case where you can receive updates on a dormant actor that isn't going to remain dormant? Or if there is ever a case where something is going to stop being dormant but receive no updates.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:160]: Talked through this offline.
